File: Forth_Interpreter/Swap.cpp
#include "Swap.hpp"

#include <memory>

bool Swap::Run () {
    if (GeneralDataForExecutors.Stack->empty()){
        throw Exception_EmptyStack();
    }

    int a = GeneralDataForExecutors.Stack->top();
    GeneralDataForExecutors.Stack->pop ();

    if (GeneralDataForExecutors.Stack->empty()){
        throw Exception_EmptyStack();
    }

    int b = GeneralDataForExecutors.Stack->top();
    GeneralDataForExecutors.Stack->pop ();
    GeneralDataForExecutors.Stack->push (a);
    GeneralDataForExecutors.Stack->push (b);

    return false;
}

File: Forth_Interpreter_On_Java/src/main/java/ru/nsu/yurakov/Forth/Swap/Swap.java
package ru.nsu.yurakov.Forth.Swap;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ru.nsu.yurakov.Forth.ForthCommands.ForthCommands;
import ru.nsu.yurakov.Forth.ForthRuntimeException.ForthRuntimeException;

/**
 * Swaps the top and second numbers on the stack
 */
public class Swap extends ForthCommands{

    @Override
    public boolean Run() throws ForthRuntimeException {
        LOGGER.info("Was launched " + this.getClass().getName());
        int a = 0;
        int b = 0;
        try {
            a = GeneralDataForExecutors.stack.peek();
            GeneralDataForExecutors.stack.pop();
        } catch (RuntimeException ex) {
            throw new ForthRuntimeException("Stack is empty");
        }
        try {
            b = GeneralDataForExecutors.stack.peek ();
            GeneralDataForExecutors.stack.pop();
        } catch (RuntimeException ex) {
            GeneralDataForExecutors.stack.push(a);
            throw new ForthRuntimeException("Stack is empty");
        }

        GeneralDataForExecutors.stack.push (a);
        GeneralDataForExecutors.stack.push (b);

        return false;
    }

    private static final Logger LOGGER = LogManager.getLogger(Swap.class);
}
